File: scraper/scraper.js
const rp = require('request-promise');
const cheerio = require('cheerio');

const { baseUrl, query, itemsPerPage, ratingsMap, selectors, regEx, timeout } = require('./utils')

class Scraper {
  constructor() {
    this.rateLimiterDelay = 1500;
    this.restaurants = [];
    this.totalPages = 1;
  }

  run = async () => {
    await this.scrape(1);
  }

  getRestaurants = () => this.restaurants;

  removeDuplicates = () => {
    this.restaurants = this.restaurants.filter((r, index, self) =>
      index === self.findIndex((t) => JSON.stringify(t) === JSON.stringify(r))
    )
  }

  setTotalPages = $ => {
    const totalItemsText = $('h1').text();
    const match = totalItemsText.match(regEx.totalItems);
  
    if (match && match.length > 0) {
      const totalItems = Number(match[0].replace(/[\.,]/g, ''));
      this.totalPages = Math.ceil(totalItems / itemsPerPage);
      console.log("Total pages to scrape:", this.totalPages);
    } else {
      console.log("Failed to find total items in the page", totalItemsText);
      this.totalPages = 0; 
    }
  }
  
  scrape = async (currentPage) => {
    const that = this;
    const start = new Date();
    const options = {
      uri: baseUrl + query + currentPage,
      transform: body => {
        return cheerio.load(body);
      }
    };

    const $ = await rp(options).catch((err) => {
      delete err.response
      console.log(err)
    });

    if (this.totalPages === 1) {
      this.setTotalPages($);
    }

    // due to errors try to collect cards if fail then simply continue
    try {
      const cards = $(selectors.cards);
      cards.each(function () {
        const $card = $(this)
        const rating = $card.find(selectors.rating).html().match(/michelin-award/g).length;
        const year = $card.find(selectors.year).text().trim();
        const img = $card.find(selectors.img).html().match(regEx.img)[0];
        const $title = $card.find(selectors.name);
        const name = $title.text().trim();
        const link = `${baseUrl}${$title.attr('href')}`;
        const location = $card.find(selectors.location).attr('data-dtm-city').trim();
        const type = $card.find(selectors.type).attr('data-cooking-type').trim();
        const lat = $card.data('lat');
        const lng = $card.data('lng');

        that.restaurants.push({ rating, year, img, name, link, location, type, lat, lng });
      });
    }
    catch (err) {
      console.log(err)
    }

    const end = new Date();
    const delay = this.rateLimiterDelay - (end - start)
    await timeout(delay)

    if (currentPage < this.totalPages) {
      return this.scrape(currentPage + 1);
    }
    return;
  }
}

module.exports = Scraper;

File: scraper/scraper.py
import requests
from bs4 import BeautifulSoup as bs
from restaurant import Restaurant
import time

base_url = 'https://guide.michelin.com/'
query = 'en/restaurants/3-stars-michelin/2-stars-michelin/1-star-michelin/page/'
cards_per_page = 20
ratings_map = {'o': 3, 'n': 2, 'm': 1}

selectors = {
    'cards': '.card__menu',
    'rating': '.fa-michelin',
    'guide': '.card__menu-content--rating>span',
    'img': '.card__menu-image>a>noscript>img',
    'name': '.card__menu-content--title>a',
    'link': '.card__menu-content--title>a',
    'location': '.card__menu-footer--location',
    'type': '.card__menu-footer--price',
    'page_numbers': '.search-results__btn-carousel-wrapper>div>ul>li>a'
}


class Scraper:
    def __init__(self):
        self.restaurants = []
        self.url = base_url + query + '1'
        self.content = bs(requests.get(self.url).content, 'html.parser')
        self.delay_rate = 0.15
        self.total_pages = self.get_total_pages(self.content)

    def get_restaurants(self):
        return self.restaurants

    def update_content(self, current_page):
        self.url = base_url + query + str(current_page)
        self.content = bs(requests.get(self.url).content, 'html.parser')

    def remove_duplicates(self):
        self.restaurants = [
            i for n, i in enumerate(self.restaurants)
            if i not in self.restaurants[:n]
        ]

    def get_total_pages(self, content):
        page_numbers = content.select(selectors["page_numbers"])[-2].get_text()
        return int(page_numbers)

    def scrape(self, current_page=1):
        self.update_content(current_page)
        cards = self.content.select(selectors['cards'])

        print("Scraping:", self.url)

        for card in cards:
            rating_tag = card.select_one(selectors['rating']).get_text()
            rating = ratings_map[rating_tag]

            guide = card.select_one(selectors['guide']).get_text().strip()

            img_tag = card.select_one(selectors['img'])
            img = img_tag.attrs["src"].split("?", 1)[0]

            name_tag = card.select_one(selectors['name'])
            name = name_tag.get_text().strip()

            link = base_url + name_tag.attrs["href"]
            location = card.select_one(
                selectors['location']).get_text().strip()
            type = card.select_one(selectors['type']).get_text().strip()

            lat = card.attrs["data-lat"]
            long = card.attrs["data-lng"]

            cur_restaurant = Restaurant(name, rating, guide, img, link,
                                        location, type, lat, long)
            self.restaurants.append(cur_restaurant)

        time.sleep(self.delay_rate)

        if current_page < self.total_pages:
            return self.scrape(current_page + 1)
